esp-hal/src/spi/master.rs

```diff
@@ -756,75 +654,37 @@ where
     pub fn with_sio3<SIO3: PeripheralOutput + PeripheralInput>(
         self,
         sio3: impl Peripheral<P = SIO3> + 'd,
-    ) -> Self {
+    ) -> Self
+    where
+        T: QspiInstance,
+    {
         crate::into_ref!(sio3);
         sio3.enable_input(true, private::Internal);
         sio3.enable_output(true, private::Internal);
 
-        sio3.connect_input_to_peripheral(self.spi.sio3_input_signal(), private::Internal);
-        sio3.connect_peripheral_to_output(self.spi.sio3_output_signal(), private::Internal);
-
-        self
-    }
-
-    /// Setup pins for this SPI instance.
-    ///
-    /// All pins are optional. Pass [crate::gpio::NoPin] if you don't need the
-    /// given pin.
-    pub fn with_pins<
-        SCK: PeripheralOutput,
-        MOSI: PeripheralOutput + PeripheralInput,
-        MISO: PeripheralOutput + PeripheralInput,
-        SIO2: PeripheralOutput + PeripheralInput,
-        SIO3: PeripheralOutput + PeripheralInput,
-        CS: PeripheralOutput,
-    >(
-        self,
-        sck: impl Peripheral<P = SCK> + 'd,
-        mosi: impl Peripheral<P = MOSI> + 'd,
-        miso: impl Peripheral<P = MISO> + 'd,
-        sio2: impl Peripheral<P = SIO2> + 'd,
-        sio3: impl Peripheral<P = SIO3> + 'd,
-        cs: impl Peripheral<P = CS> + 'd,
-    ) -> Self {
-        self.with_sck(sck)
-            .with_mosi(mosi)
-            .with_miso(miso)
-            .with_sio2(sio2)
-            .with_sio3(sio3)
-            .with_cs(cs)
-    }
-    /// Change the bus frequency of the SPI instance in half-duplex mode.
-    ///
-    /// This method allows you to update the bus frequency for the SPI
-    /// communication after the instance has been created.
-    pub fn change_bus_frequency(&mut self, frequency: HertzU32) {
-        self.spi.ch_bus_freq(frequency);
-    }
+        sio3.connect_input_to_peripheral(unwrap!(self.spi.sio3_input_signal()), private::Internal);
+        sio3.connect_peripheral_to_output(
+            unwrap!(self.spi.sio3_output_signal()),
+            private::Internal,
+        );
 
-    /// Set the bit order for the SPI instance.
-    ///
-    /// The default is MSB first for both read and write.
-    pub fn with_bit_order(mut self, read_order: SpiBitOrder, write_order: SpiBitOrder) -> Self {
-        self.spi.set_bit_order(read_order, write_order);
         self
     }
 }
 
-impl<T> HalfDuplexReadWrite for Spi<'_, HalfDuplexMode, T>
+impl<T> Spi<'_, T>
 where
     T: Instance,
 {
-    type Error = Error;
-
-    fn read(
+    /// Half-duplex read.
+    pub fn half_duplex_read(
```

Comment:
I'm guessing you may have had some internal discussions about this, what were the thoughts around removing the "full duplex" write and read, to reduce the API surface?

So the only available methods would be:
- write (half duplex)
- read (half duplex)
- transfer (full duplex)

Reply:
It would be very confusing for me, and I think it would be very confusing for our users. The half-duplex functions are pretty hardware-specific, and pretty advanced, and they are not working as a "write but ignore read data" and "read without me giving you something to write". If someone would force their way through `write` and `read` thinking it's doing exactly that, they will not understand what's going on.

Similar looking function names (i.e., in this case a single verb without any qualifier) suggest similar functionality. A `half_duplex_` prefix IMHO makes these functions distinctive enough to make the user think about them, and whether they want them.